File: ugs-status-publisher-server/src/main/java/no/tjer/teamcity/ugsStatusPublisher/UgsStatusPublisher.java
/*
 * Copyright 2000-2022 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package no.tjer.teamcity.ugsStatusPublisher;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;

import java.io.IOException;
import java.security.KeyStore;
import java.text.MessageFormat;
import java.util.*;

import jetbrains.buildServer.commitPublisher.*;
import jetbrains.buildServer.log.LogUtil;
import jetbrains.buildServer.messages.Status;
import jetbrains.buildServer.serverSide.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;

import static jetbrains.buildServer.commitPublisher.LoggerUtil.LOG;


/**
 * Updates TFS Git commit statuses via REST API.
 */
class UgsStatusPublisher extends HttpBasedCommitStatusPublisher<UgsStatusPublisher.BadgeResult> {

    private static final String METRICS_URL_FORMAT = "{0}/api/rugs_metrics";
    private static final String POST_BADGE_URL_FORMAT = "{0}/api/build";
    private static final String ERROR_AUTHORIZATION = "Check username & password for RUGS";
    private static final String FAILED_TO_TEST_CONNECTION_TO_REPOSITORY = "UGS publisher has failed to test connection to server ";
    private static final Gson myGson = new GsonBuilder()
            .setDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
            .create();

    UgsStatusPublisher(@NotNull final CommitStatusPublisherSettings settings,
                       @NotNull final SBuildType buildType,
                       @NotNull final String buildFeatureId,
                       @NotNull final WebLinks webLinks,
                       @NotNull final Map<String, String> params,
                       @NotNull final CommitStatusPublisherProblems problems) {
        super(settings, buildType, buildFeatureId, params, problems, webLinks);
    }

    @NotNull
    public String toString() {
        return UgsConstants.ID;
    }

    @NotNull
    @Override
    public String getId() {
        return UgsConstants.ID;
    }

    @Override
    public boolean isPublishingForRevision(@NotNull final BuildRevision revision) {
        return true;
    }

    @Override
    public boolean buildStarted(@NotNull SBuild build, @NotNull BuildRevision revision) throws PublisherException {
        updateBuildStatus(build, revision, true);
        return true;
    }

    @Override
    public boolean buildFinished(@NotNull SBuild build, @NotNull BuildRevision revision) throws PublisherException {
        updateBuildStatus(build, revision, false);
        return true;
    }

    @Override
    public boolean buildInterrupted(@NotNull SBuild build, @NotNull BuildRevision revision) throws PublisherException {
        updateBuildStatus(build, revision, false);
        return true;
    }

    @Override
    public boolean buildMarkedAsSuccessful(@NotNull final SBuild build, @NotNull final BuildRevision revision, final boolean buildInProgress) throws PublisherException {
        updateBuildStatus(build, revision, buildInProgress);
        return true;
    }

    @Override
    public void processResponse(@NotNull final HttpHelper.HttpResponse response) throws HttpPublisherException, IOException {
        if (response.getStatusCode() >= 400) {
            processErrorResponse(response);
        }
    }

    @Override
    public RevisionStatus getRevisionStatus(@NotNull BuildPromotion buildPromotion, @NotNull BuildRevision revision) throws PublisherException {
        return null;  // TODO?
    }

    public static void testConnection(@NotNull final Map<String, String> params,
                                      @Nullable final KeyStore trustStore) throws PublisherException {

        String serverUrl = params.get(UgsConstants.SERVER_URL);
        String url = MessageFormat.format(METRICS_URL_FORMAT, serverUrl);
        try {
            IOGuard.allowNetworkCall(() -> {
                HttpHelper.get(url, getUserName(params), getPassword(params), Collections.singletonMap("Accept", "application/json"), BaseCommitStatusPublisher.DEFAULT_CONNECTION_TIMEOUT,
                        trustStore, new DefaultHttpResponseProcessor() {
                            @Override
                            public void processResponse(HttpHelper.HttpResponse response) throws HttpPublisherException, IOException {
                                final int status = response.getStatusCode();
                                if (status == 401 || status == 403) {
                                    throw new HttpPublisherException(ERROR_AUTHORIZATION);
                                }

                                // Ignore Bad Request for POST check
                                if (status == 400) {
                                    return;
                                }

                                if (status != 200) {
                                    processErrorResponse(response);
                                }
                            }
                        });
            });

        } catch (Exception e) {
            final String message = FAILED_TO_TEST_CONNECTION_TO_REPOSITORY + serverUrl;
            LOG.debug(message, e);
            throw new PublisherException(message, e);
        }
    }

    private static void processErrorResponse(@NotNull final HttpHelper.HttpResponse response) throws HttpPublisherException {
        final int status = response.getStatusCode();
        final String content = response.getContent();
        if (null == content) {
            throw new HttpPublisherException(status, response.getStatusText(), "Empty HTTP response");
        }

        Error error = null;
        try {
            error = myGson.fromJson(content, Error.class);
        } catch (JsonSyntaxException e) {
            // Invalid JSON response
        }

        final String message;
        if (error != null && error.message != null) {
            message = error.message;
        } else {
            message = "HTTP response error";
        }

        throw new HttpPublisherException(status, response.getStatusText(), message);
    }

    private void updateBuildStatus(@NotNull SBuild build, @NotNull BuildRevision revision, boolean isStarting) throws PublisherException {
        final CommitStatus status = getCommitStatus(build, isStarting);
        final String description = LogUtil.describe(build);
        final String commitStatusUrl = MessageFormat.format(POST_BADGE_URL_FORMAT, myParams.get(UgsConstants.SERVER_URL));

        JSONObject body = new JSONObject();
        // TODO: Try parse?
        body.put("ChangeNumber", Integer.parseInt(revision.getRevision(), 10));
        body.put("Project", myParams.get(UgsConstants.PROJECT));
        body.put("BuildType", myParams.get(UgsConstants.BADGE_NAME));
        body.put("Result", status.result.getUGSValue());
        body.put("Url", status.targetUrl);

        postJson(commitStatusUrl, getUserName(myParams), getPassword(myParams),
                body.toString(),
                Collections.singletonMap("Accept", "application/json"),
                description
        );
    }

    @NotNull
    private CommitStatus getCommitStatus(final SBuild build, final boolean isStarting) {
        BuildPromotion buildPromotion = build.getBuildPromotion();
        boolean isCanceled = buildPromotion.isCanceled() || build.isInternalError();
        LOG.warn(String.format("UGS %s: isStarting: %s isCanceled: %s isISE: %s Status: %s", build.getFullName(), isStarting, buildPromotion.isCanceled(), build.isInternalError(), build.getBuildStatus()));
        BadgeResult badge = getBadge(isStarting, isCanceled, build.getBuildStatus());
        return new CommitStatus(badge, getViewUrl(build));
    }

    private static BadgeResult getBadge(boolean isStarting, boolean isCanceled, Status status) {
        if (!isStarting) {
            if (status.isSuccessful()) return BadgeResult.SUCCESS;
            else if (status == Status.ERROR) return BadgeResult.FAILURE;
            else if (status == Status.FAILURE) return BadgeResult.FAILURE;
            else if (isCanceled) return BadgeResult.SKIPPED;
            else if (status == Status.WARNING) return BadgeResult.WARNING;
            else {
                LOG.warn("Unknown status " + status);
                return BadgeResult.SKIPPED;
            }
        }

        return BadgeResult.STARTING;
    }

    @Nullable
    private static String getUserName(Map<String, String> params) {
        final String user = params.get(UgsConstants.AUTH_USER);
        final String password = params.get(UgsConstants.AUTH_PASSWORD);
        return password != null ? user : null;
    }

    @Nullable
    private static String getPassword(Map<String, String> params) {
        return params.get(UgsConstants.AUTH_PASSWORD);
    }

    private static class Error {
        String message;
    }

    static class CommitStatus {
        @NotNull
        final BadgeResult result;
        final String targetUrl;

        public CommitStatus(@NotNull BadgeResult result, String targetUrl) {
            this.result = result;
            this.targetUrl = targetUrl;
        }
    }

    enum BadgeResult {
        STARTING("Starting", 0),
        FAILURE("Failure", 1),
        WARNING("Warning", 2),
        SUCCESS("Success", 3),
        SKIPPED("Skipped", 4);

        private final String displayName;
        private final int ugsValue;

        BadgeResult(String displayName, int ugsValue) {
            this.displayName = displayName;
            this.ugsValue = ugsValue;
        }

        public String getDisplayName() {
            return displayName;
        }

        public int getUGSValue() {
            return ugsValue;
        }
    }
}

File: src/main/java/no/tjer/teamcity/ugsStatusPublisher/UgsStatusPublisher.kt
/*
 * Copyright 2023 Jørgen Tjernø <jorgen@tjer.no>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package no.tjer.teamcity.ugsStatusPublisher

import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import jetbrains.buildServer.commitPublisher.*
import jetbrains.buildServer.log.LogUtil
import jetbrains.buildServer.messages.Status
import jetbrains.buildServer.serverSide.*
import jetbrains.buildServer.vcshostings.http.HttpHelper
import jetbrains.buildServer.vcshostings.http.credentials.HttpCredentials
import jetbrains.buildServer.vcshostings.http.credentials.UsernamePasswordCredentials
import no.tjer.teamcity.ugsStatusPublisher.UgsStatusPublisher.BadgeResult
import org.json.JSONObject
import java.io.IOException
import java.security.KeyStore
import java.text.MessageFormat
import java.util.*

val BuildPromotion.isCanceledOrFailedToStart: Boolean
    get() {
        if (isCanceled) {
            return true;
        }
        return associatedBuild?.isInternalError ?: false
    }

/**
 * Updates Unreal Game Sync commit statuses via REST API.
 */
internal class UgsStatusPublisher(
    settings: CommitStatusPublisherSettings,
    buildType: SBuildType,
    buildFeatureId: String,
    webLinks: WebLinks,
    params: Map<String, String>,
    problems: CommitStatusPublisherProblems
) : HttpBasedCommitStatusPublisher<BadgeResult?>(settings, buildType, buildFeatureId, params, problems, webLinks) {
    override fun toString(): String {
        return UgsConstants.ID
    }

    override fun getId(): String {
        return UgsConstants.ID
    }

    override fun isPublishingForRevision(revision: BuildRevision): Boolean {
        return true
    }

    @Throws(PublisherException::class)
    override fun buildStarted(build: SBuild, revision: BuildRevision): Boolean {
        updateBuildStatus(build, revision, true)
        return true
    }

    @Throws(PublisherException::class)
    override fun buildFinished(build: SBuild, revision: BuildRevision): Boolean {
        updateBuildStatus(build, revision, false)
        return true
    }

    @Throws(PublisherException::class)
    override fun buildInterrupted(build: SBuild, revision: BuildRevision): Boolean {
        updateBuildStatus(build, revision, false)
        return true
    }

    @Throws(PublisherException::class)
    override fun buildMarkedAsSuccessful(build: SBuild, revision: BuildRevision, buildInProgress: Boolean): Boolean {
        updateBuildStatus(build, revision, buildInProgress)
        return true
    }

    @Throws(HttpPublisherException::class, IOException::class)
    override fun processResponse(response: HttpHelper.HttpResponse) {
        if (response.statusCode >= 400) {
            processErrorResponse(response)
        }
    }

    @Throws(PublisherException::class)
    override fun getRevisionStatus(buildPromotion: BuildPromotion, revision: BuildRevision): RevisionStatus? {
        return null // TODO?
    }

    @Throws(PublisherException::class)
    private fun updateBuildStatus(build: SBuild, revision: BuildRevision, isStarting: Boolean) {
        val status = getCommitStatus(build, isStarting)
        val description = LogUtil.describe(build)
        val commitStatusUrl = MessageFormat.format(POST_BADGE_URL_FORMAT, myParams[UgsConstants.SERVER_URL])
        val body = JSONObject()
        // TODO: Try parse?

        // This will return the changelist number if using feature branches (e.g. "main|123" -> 123)
        // or only the changelist number if not using feature branches ("123" -> 123)
        val changeNumber = revision.revision.split("|").last().toInt(10)
        body.put("ChangeNumber", changeNumber)
        body.put("Project", myParams[UgsConstants.PROJECT])
        body.put("BuildType", myParams[UgsConstants.BADGE_NAME])
        body.put("Result", status.result.uGSValue)
        body.put("Url", status.targetUrl)
        postJson(
            commitStatusUrl,
            getCredentials(myParams),
            body.toString(),
            Collections.singletonMap("Accept", "application/json"),
            description
        )
    }

    private fun getCommitStatus(build: SBuild, isStarting: Boolean): CommitStatus {
        val buildPromotion = build.buildPromotion
        // If a dependency has failed for a composite build, report a failure even if we're canceled
        val dependencies = buildPromotion.dependencies.map { it.dependOn }
        if (build.isCompositeBuild) {
            if (dependencies.any { !it.isCanceledOrFailedToStart && it.associatedBuild?.buildStatus?.isFailed == true }) {
                return CommitStatus(BadgeResult.FAILURE, getViewUrl(build))
            }
        }

        val isCanceled = buildPromotion.isCanceled || build.isInternalError
        val badge = getBadge(isStarting, isCanceled, build.buildStatus)
        return CommitStatus(badge, getViewUrl(build))
    }

    private class Error {
        var message: String? = null
    }

    internal class CommitStatus(val result: BadgeResult, val targetUrl: String)
    internal enum class BadgeResult(val displayName: String, val uGSValue: Int) {
        STARTING("Starting", 0), FAILURE("Failure", 1), WARNING("Warning", 2), SUCCESS("Success", 3), SKIPPED(
            "Skipped", 4
        )
    }

    companion object {
        private const val METRICS_URL_FORMAT = "{0}/api/rugs_metrics"
        private const val POST_BADGE_URL_FORMAT = "{0}/api/build"
        private const val ERROR_AUTHORIZATION = "Check username & password for RUGS"
        private const val FAILED_TO_TEST_CONNECTION_TO_REPOSITORY =
            "UGS publisher has failed to test connection to server "
        private val myGson = GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").create()

        @Throws(PublisherException::class)
        fun testConnection(
            params: Map<String, String>, trustStore: KeyStore?
        ) {
            val serverUrl = params[UgsConstants.SERVER_URL]
            val url = MessageFormat.format(METRICS_URL_FORMAT, serverUrl)
            try {
                IOGuard.allowNetworkCall<Exception> {
                    HttpHelper.get<HttpPublisherException>(url,
                        getCredentials(params),
                        Collections.singletonMap<String, String>("Accept", "application/json"),
                        DEFAULT_CONNECTION_TIMEOUT,
                        trustStore,
                        object : DefaultHttpResponseProcessor() {
                            @Throws(HttpPublisherException::class, IOException::class)
                            override fun processResponse(response: HttpHelper.HttpResponse) {
                                val status = response.statusCode
                                if (status == 401 || status == 403) {
                                    throw HttpPublisherException(ERROR_AUTHORIZATION)
                                }

                                // Ignore Bad Request for POST check
                                if (status == 400) {
                                    return
                                }
                                if (status != 200) {
                                    processErrorResponse(response)
                                }
                            }
                        })
                }
            } catch (e: Exception) {
                val message = FAILED_TO_TEST_CONNECTION_TO_REPOSITORY + serverUrl
                LoggerUtil.LOG.debug(message, e)
                throw PublisherException(message, e)
            }
        }

        @Throws(HttpPublisherException::class)
        private fun processErrorResponse(response: HttpHelper.HttpResponse) {
            val status = response.statusCode
            val content =
                response.content ?: throw HttpPublisherException(status, response.statusText, "Empty HTTP response")
            var error: Error? = null
            try {
                error = myGson.fromJson(content, Error::class.java)
            } catch (e: JsonSyntaxException) {
                // Invalid JSON response
            }
            val message = error?.message ?: "HTTP response error"
            throw HttpPublisherException(status, response.statusText, message)
        }

        private fun getBadge(isStarting: Boolean, isCanceled: Boolean, status: Status): BadgeResult {
            return if (!isStarting) {
                if (status.isSuccessful) {
                    BadgeResult.SUCCESS
                } else if (isCanceled) {
                    BadgeResult.SKIPPED
                } else if (status == Status.ERROR) {
                    BadgeResult.FAILURE
                } else if (status == Status.FAILURE) {
                    BadgeResult.FAILURE
                } else if (status == Status.WARNING) {
                    BadgeResult.WARNING
                } else {
                    LoggerUtil.LOG.warn("Unknown status $status")
                    BadgeResult.SKIPPED
                }
            } else BadgeResult.STARTING
        }

        private fun getCredentials(params: Map<String, String>): HttpCredentials? {
            val user = params[UgsConstants.AUTH_USER]
            val password = params[UgsConstants.AUTH_PASSWORD]
            return if (user != null && password != null) {
                UsernamePasswordCredentials(user, password)
            } else {
                null
            }
        }
    }
}
